idaes/apps/caprese/estimator.py

```diff
@@ -0,0 +1,597 @@
+        non_initial_time = list(time)[1:]
+        deactivated = deactivate_model_at(
+                model,
+                time,
+                non_initial_time,
+                allow_skip=True,
+                suppress_warnings=True,
+                )
+
+        self.vectors.differential[:, t0].unfix()
+        self.vectors.input[:, t0].unfix() #dof
+        self.vectors.derivative[:, t0].fix(0.)
+        # Solving for steady state doesn't need MHE block.
+        self.MHE_VARS_CONS_BLOCK.deactivate()
+        # Activate the original/undisturbed differential equations at t0
+        for indexcon in self.con_category_dict[CC.DIFFERENTIAL]:
+            indexcon[t0].activate()
+
+        self.steadystate_objective.activate()
+
+        dof = degrees_of_freedom(model)
+        # This should be True for solving a steady state.
+        assert dof == len(self.INPUT_SET)
+
+        results = solver.solve(self, tee=True)
+        if results.solver.termination_condition == TerminationCondition.optimal:
+            pass
+        else:
+            msg = 'Failed to solve for full steady state values'
+            raise RuntimeError(msg)
+
+        self.steadystate_objective.deactivate()
+
+        for indexcon in self.con_category_dict[CC.DIFFERENTIAL]:
+            indexcon[t0].deactivate()
+        self.MHE_VARS_CONS_BLOCK.activate()
+        self.vectors.derivative[:, t0].unfix()
+        self.vectors.input[:, t0].fix()
+        self.vectors.differential[:, t0].fix()
+
+        for t, complist in deactivated.items():
+            for comp in complist:
+                if was_originally_active[comp]:
+                    comp.activate()
+
+    def initialize_actualmeasurements_at_t0(self):
+        t0 = self.sample_points[0]
+        for ind in self.MEASUREMENT_SET:
+            actmea_block = self.ACTUALMEASUREMENT_BLOCK[ind]
+            mea_block = self.MEASUREMENT_BLOCK[ind]
+            actmea_block.var[0].set_value(mea_block.var[0].value)  
+
+    def initialize_past_info_with_steady_state(self, 
+                                               desired_ss, 
+                                               ss_weights, 
+                                               solver):
+        '''
+        Before running MHE, we need to set up past information (measurements, states, etc.). 
+        Here we solve for a steady state and set it as past information.
+
+        Parameters
+        ----------
+        desired_ss : List of vardata, value tuples describing the
+                     desired steady stsates of these specified variables.
+        ss_weights : List of vardata, value tuples describing the
+                     weightss of these specified variables.
+
+        '''
+        self.add_steady_state_objective(desired_ss, ss_weights)
+        self.solve_steady_state(solver)
+
+        self.initialize_actualmeasurements_at_t0()
+        self.initialize_to_initial_conditions(ctype=(DiffVar, AlgVar, DerivVar, InputVar,))
```

Comment:
I have to say I put those together is based on the fact that we use the steady state to initialize the past info.

Which of these is "initializing past info"?
Ans: Yes, both. Once I get the steady state, I first update the value of `actualmeasurement` at `t0`. Then I take advantage of `initialize_to_initial_conditions` to initialize variables at the other time points.

I would think that the "actualmeasurement vars" are the "past info", and the initialization of the estimator model should happen separately.
Ans: Well, I would say it depends on how much "past info" do you have. Of course, past "actualmeasurement vars" and "control inputs" are necessary, but if you have past "estimates", you can use those to initialize state variables. I understand you may want to initialize the estimator model separately, and I think it's related to the previous comment about how we initialize the past info. Let's discuss how to put these together in a systematic way. Good point!

Are the actualmeasurement vars initialized other than at t0? Does this need to happen?
Ans: Yes, it's initialized in function `initialize_to_initial_conditions`, but I just found I included "actualmeasurement" in a stupid way, (in `initialize_to_initial_conditions`)
```
if ActualMeasurementVar in self.collect_ctypes():
    ctype += (ActualMeasurementVar,)
```
I've changed it to the following, 
```
self.initialize_to_initial_conditions(ctype=(DiffVar, AlgVar, DerivVar, InputVar, ActualMeasurementVar))
```

Reply:
Okay, I think I see what you're saying about both actualmeasurement vars and estimator model vars being "past info." I'm trying to decide if they are different enough to warrant being in separate functions. I guess the main difference is that actualmeasurement vars only exist at sample points, where the model's variables exist everywhere. 

However, I think the main challenge here is that we may want to initialize either of these from time-varying data from a plant, simulation, or estimation, which we don't currently have a good way to do. Like (I think) I've said elsewhere, I'm a little hesitant to have an embedded solve here, but I think I'm okay with initializing all the variables from a single model in this function. I think we should also support initializing actualmeasurements separately, and initialing from somewhat arbitrary time-indexed data. This can wait until the next PR, I think.